core/src/kmx/kmx_plus.cpp

```diff
@@ -0,0 +1,224 @@
+    for(int j=0; str[j] && j<0x30; j++) {
+        if (str[j] < 0x7F && str[j] != 0x0020 && str[j] > 0x20) {
+            putchar(str[j]);
+        } else {
+            printf("U+%04X ", str[j]);
+        }
+    }
+    printf("\n");
+  }
+}
+
+static void
+dump_kmxplus_sect(const uint8_t* data, const COMP_KMXPLUS_SECT* sect) {
+  dump_kmxplus_header((const COMP_KMXPLUS_HEADER*)sect);
+  printf("sect: total 0x%X\n", sect->total);
+  printf("sect: count 0x%X\n", sect->count);
+
+  for (KMX_DWORD i = 0; i < sect->count; i++) {
+    const COMP_KMXPLUS_SECT_ENTRY& entry = sect->entries[i];
+    dump_section_name(entry.sect);
+    printf(" sect#%d: %X @ %X\n", i, entry.sect, entry.offset);
+    const uint8_t* entrydata = (data+entry.offset);
+    switch(entry.sect) {
+        case LDML_SECTION_KEYS:
+            dump_kmxplus_keys(data, as_kmxplus_keys(entrydata));
+            break;
+        case LDML_SECTION_LOCA:
+            dump_kmxplus_loca(data, as_kmxplus_loca(entrydata));
+            break;
+        case LDML_SECTION_META:
+            dump_kmxplus_meta(data, as_kmxplus_meta(entrydata));
+            break;
+        case LDML_SECTION_SECT:
+            printf("! Cowardly refusing to dump nested 'sect' section.\n");
+            break;
+        case LDML_SECTION_STRS:
+            dump_kmxplus_strs(data, as_kmxplus_strs(entrydata));
+            break;
+        case LDML_SECTION_VKEY:
+            dump_kmxplus_vkey(data, as_kmxplus_vkey(entrydata));
+            break;
+        default:
+            printf("Unknown section %X", entry.sect);
+    }
+  }
+}
+#endif
+
+void
+dump_kmxplus_data(const uint8_t* data) {
+#if KMXPLUS_DEBUG
+  const COMP_KMXPLUS_SECT* sect = as_kmxplus_sect(data);
+  if (sect == NULL) {
+    printf("Err: 'sect' null from %p\n", data);
+    return;
+  }
+  dump_kmxplus_sect(data, sect);
+#endif
+}
+
+void
+dump_kmxplus_data(kmx::PCOMP_KEYBOARD keyboard) {
+#if KMXPLUS_DEBUG
+  printf("dump_kmxplus_data(): Got a PCOMP_KEYBOARD at %p\n", keyboard);
+  if (!(keyboard->dwFlags & KF_KMXPLUS)) {
+    printf("Err: flags KF_KMXPLUS not set\n");
+    return;
+  }
+  const COMP_KEYBOARD_EX* ex = reinterpret_cast<const COMP_KEYBOARD_EX*>(keyboard);
+
+  printf("KMXPlus offset 0x%X, KMXPlus size 0x%X\n", ex->kmxplus.dpKMXPlus, ex->kmxplus.dwKMXPlusSize);
+  const uint8_t* rawdata = reinterpret_cast<const uint8_t*>(keyboard);
+  dump_kmxplus_data(rawdata + ex->kmxplus.dpKMXPlus);
+#endif
+}
+
+const COMP_KMXPLUS_KEYS_ENTRY *COMP_KMXPLUS_KEYS::find(KMX_DWORD vkey, KMX_DWORD mod) const {
+    // TODO-LDML: eventually, assume sorted order & binary search
+    for (KMX_DWORD i=0; i<count; i++) {
+        if(entries[i].vkey == vkey && entries[i].mod == mod) {
+            return &entries[i];
+        }
+    }
+    return NULL;
+}
+
+KMX_DWORD COMP_KMXPLUS_SECT::find(KMX_DWORD ident) const {
+  for (KMX_DWORD i = 0; i < count; i++) {
+    if (ident == entries[i].sect) {
+        return entries[i].offset;
+    }
+  }
+  return 0;
+}
```

Comment:
Alternatively, on load we could fill `std::maps` with these and leverage existing C++ classes?

Reply:
Yes. That's actually the next PR.